Fix for crash at chrome-untrusted://chat url

AlexeyBarabash opened the pull request:
<!-- Add brave-browser issue below that this PR will resolve -->
Resolves https://github.com/brave/brave-browser/issues/33995

This PR does 2 things:
1. fixes the crash
2. prevent the page chrome-untrusted://chat from being opened in tab

Crash happened because when `chrome-untrusted://chat` page opened in a regular tab and not in a panel, the `AIChatUIPageHandler::active_chat_tab_helper_` wasn't set. It won't happen with a normal use of the chat, but I have added a checks to prevent crashing if it somehow will happen.

Preventing of opening the chat in a regular tab is  done by through `transition_type` parameter.
It's value is different on Android and the desktop, see the table at the comments at `BraveNavigationControllerImpl::LoadURLWithParams`.

<!-- CI-related labels that can be applied to this PR:
* CI/run-audit-deps (1) - check for known npm/cargo vulnerabilities (audit_deps)
* CI/run-network-audit (1) - run network-audit
* CI/run-upstream-tests - run Chromium unit and browser tests on Linux and Windows (otherwise only on Linux)
* CI/run-linux-arm64, CI/run-macos-arm64, CI/run-windows-arm64, CI/run-windows-x86 - run builds that would otherwise be skipped
* CI/skip - do not run CI builds (except noplatform)
* CI/skip-linux-x64, CI/skip-android, CI/skip-macos-x64, CI/skip-ios, CI/skip-windows-x64 - skip CI builds for specific platforms
* CI/skip-upstream-tests - do not run Chromium unit, or browser tests (otherwise only on Linux)
* CI/skip-all-linters - do not run presubmit and lint checks
* CI/storybook-url (1) - deploy storybook and provide a unique URL for each build

(1) applied automatically when some files are changed (see: https://github.com/brave/brave-core/blob/master/.github/labeler.yml)
-->

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
See https://github.com/brave/brave-browser/issues/33995

thypon commented:
sec gtg

cdesouza-chromium commented on content/browser/renderer_host/brave_navigation_controller_impl.h:
```diff
@@ -0,0 +1,36 @@
+class FrameTree;
+class NavigationControllerDelegate;
+
+class CONTENT_EXPORT BraveNavigationControllerImpl
```
Do you think it would help to have some comment above explaining the role of `BraveNavigationControllerImpl`, as opposed to the class it is overrding.

AlexeyBarabash replied:
thanks, fixed at https://github.com/brave/brave-core/pull/20996/commits/f840e5dc3a8a356859d6cd721ee8345ad7f19a85#diff-22e6afdf9bd6b125060d65f811409fd5c8a4a3b72b7c48ab946d83adc76bc90fR17

cdesouza-chromium commented on content/browser/renderer_host/brave_navigation_controller_impl.cc:
```diff
@@ -0,0 +1,44 @@
+  if (params.url.SchemeIs(content::kChromeUIUntrustedScheme) &&
+      params.url.host() == kChatUIHost) {
+#if BUILDFLAG(IS_ANDROID)
+    if (!ui::PageTransitionTypeIncludingQualifiersIs(
```
I feel it would be helpful to have an explanation why android code is handled differently here.

AlexeyBarabash replied:
thanks, fixed at https://github.com/brave/brave-core/pull/20996/commits/f840e5dc3a8a356859d6cd721ee8345ad7f19a85#diff-fc1feda162362111f94891cf8248d4682440d13c393d7cdb1e600178ea16b230R24

cdesouza-chromium commented on content/browser/renderer_host/brave_navigation_controller_impl.cc:
```diff
@@ -0,0 +1,44 @@
+base::WeakPtr<NavigationHandle>
+BraveNavigationControllerImpl::LoadURLWithParams(const LoadURLParams& params) {
+  if (params.url.SchemeIs(content::kChromeUIUntrustedScheme) &&
+      params.url.host() == kChatUIHost) {
```
Should there be a browser tests for this?

AlexeyBarabash replied:
Yes, it would be better, fixed at 
https://github.com/brave/brave-core/pull/20996/commits/29c81aed8b7bf633dcae5b9fab10aa04c4a46105

cdesouza-chromium commented:
I feel either PR description, or the commit message, should have some more details about the nature of the crash, and the fix for it, that can be gleaned on by someone else in the future having to look on git history.

AlexeyBarabash commented on browser/ui/webui/ai_chat/ai_chat_ui_page_handler.cc:
```diff
@@ -165,6 +171,12 @@ void AIChatUIPageHandler::SetAutoGenerateQuestions(bool value) {
 void AIChatUIPageHandler::GetSiteInfo(GetSiteInfoCallback callback) {
+  if (!active_chat_tab_helper_) {
+    VLOG(2) << "Chat tab helper is not set";
+    std::move(callback).Run(false, nullptr);
```
fixed at https://github.com/brave/brave-core/pull/20996/commits/8f478efd68cd23b3b09a9cb6f9f9ce102b1ff4ed, thanks

AlexeyBarabash commented:
@cdesouza-chromium 

> I feel either the PR description, or the commit message, should have some more details about the nature of the crash, and the fix for it, that can be gleaned on by someone else in the future having to look on git history.

I agree, I added the comments in the code and modified the PR description.

cdesouza-chromium reviewed (approved):
LGTM. Thanks for your time addressing those comments.

SergeyZhukovsky reviewed (approved):
++

goodov commented on content/browser/renderer_host/brave_navigation_controller_impl.cc:
```diff
@@ -0,0 +1,62 @@
+BraveNavigationControllerImpl::~BraveNavigationControllerImpl() {}
+
+base::WeakPtr<NavigationHandle>
+BraveNavigationControllerImpl::LoadURLWithParams(const LoadURLParams& params) {
```
do we really need to create `BraveNavigationControllerImpl` for the sole purpose of preventing `chrome-untrusted://chat` opening? To me this looks like a wrong approach. This way we had to block/ban all other unwanted pages, but we don't do that.

Maybe we can fix the chat WebUI to be no-op if it's opened in an unsupported context?

AlexeyBarabash replied:
>do we really need to create BraveNavigationControllerImpl for the sole purpose of preventing chrome-untrusted://chat opening?

this is the best place I could find for such preventing. 
I was surprised by the fact there were no any brave-related overrides or patches around `NavigationControllerImpl`, but nevertheless I had put the change there.

>Maybe we can fix the chat WebUI to be no-op if it's opened in an unsupported context?

Basically the first commit afeef6f90d9028b5e007c4082d9d849d6189afaa in the PR  fixes the crash itself at WebUI, but doesn't prevent the page from load.
I will recheck if the page transition parameter is available at this place, but as far as I remember, it doesn't.


Thanks for the feedback.

AlexeyBarabash replied:
I may try to use 
[WebContentsObserver](https://github.com/brave/brave-core/blob/master/browser/ui/webui/ai_chat/ai_chat_ui_page_handler.h#L79)
[DidStartNavigation](https://source.chromium.org/chromium/chromium/src/+/main:content/public/browser/web_contents_observer.h;drc=42f0b9c4f60f67676d9096b3e94df9fd47f568f4;bpv=1;bpt=1;l=307?q=WebContentsObserver&ss=chromium%2Fchromium%2Fsrc)
[GetPageTransition](https://source.chromium.org/chromium/chromium/src/+/main:content/public/browser/navigation_handle.h;l=269;drc=445f8cfdbb63b29ce44f728f6b5015cf18df5534)

to detect the unwanted navigations and then find the way to prevent them.

AlexeyBarabash replied:
Thanks, changed the PR to use `NavigationThrottle`
https://github.com/brave/brave-core/pull/20996/commits/025abf5e0bfe6df6297eb4de3f109f839add8239

goodov replied:
> Thanks, changed the PR to use `NavigationThrottle` [025abf5](https://github.com/brave/brave-core/commit/025abf5e0bfe6df6297eb4de3f109f839add8239)

much cleaner, thanks!

AlexeyBarabash commented on content/browser/BUILD.gn:
```diff
@@ -0,0 +1,25 @@
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at https://mozilla.org/MPL/2.0/.
+
+source_set("browser_tests") {
```
That file is gone now completely.

Can I use `chrome` from `components/xxx/content/browser` as https://github.com/brave/brave-core/pull/20996/commits/025abf5e0bfe6df6297eb4de3f109f839add8239#diff-b835d3feb3f8584d7ee0f625cca0adab7167f18ef4c19b8d4a1bf5e02e2b4313R54 ?

goodov replied:
> Can I use `chrome` from `components/xxx/content/browser` as

please don't do that. We should migrate all chrome-requiring tests from `components` into `chrome` at some point.

AlexeyBarabash replied:
@goodov does this change look ok https://github.com/brave/brave-core/pull/20996/commits/0ce243326cc3bf3d6680602112625ff6d1fd0a96 ?

The test is in `brave/browser` but the class is at `brave/components/ai_chat/content/browser`.

goodov commented on components/ai_chat/content/browser/ai_chat_throttle.h:
```diff
@@ -0,0 +1,35 @@
+
+#include <memory>
+
+#include "content/public/browser/navigation_throttle.h"
```
forward-decl or header? choose one :)

AlexeyBarabash replied:
thanks, fixed at https://github.com/brave/brave-core/pull/20996/commits/a12d71654dc6939c2ef876ee1dcf39bb7c72c569 :+1:

goodov commented on components/ai_chat/content/browser/ai_chat_throttle.cc:
```diff
@@ -0,0 +1,79 @@
+  // We need this throttle to work only for chrome-untrusted://chat page
+  if (!navigation_handle->GetURL().SchemeIs(
+          content::kChromeUIUntrustedScheme) ||
+      navigation_handle->GetURL().host() != kChatUIHost) {
```
`host_piece()`

AlexeyBarabash replied:
fixed at https://github.com/brave/brave-core/pull/20996/commits/511e52d57ee3fcbefb1b844aa289dea341305e65 :+1:

goodov reviewed (approved):
lgtm with few nits

goodov commented on browser/ai_chat/BUILD.gn:
```diff
@@ -0,0 +1,22 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at https://mozilla.org/MPL/2.0/.
+
```
```
import("//brave/components/ai_chat/core/common/buildflags/buildflags.gni")

assert(enable_ai_chat)
```

AlexeyBarabash replied:
thanks, fixed at https://github.com/brave/brave-core/pull/20996/commits/9b80721e3a8b5188bb8deb3319a1bf94543f5344

goodov commented on test/BUILD.gn:
```diff
@@ -171,6 +171,7 @@ test("brave_unit_tests") {
     "//brave/brave_domains:unit_tests",
     "//brave/browser",
     "//brave/browser:unit_tests",
+    "//brave/browser/ai_chat:unit_tests",
```
`if (enable_ai_chat)`

AlexeyBarabash replied:
thanks, fixed at https://github.com/brave/brave-core/pull/20996/commits/9b80721e3a8b5188bb8deb3319a1bf94543f5344

goodov commented on browser/brave_content_browser_client.cc:
```diff
@@ -35,6 +35,7 @@
 #include "brave/browser/profiles/profile_util.h"
 #include "brave/browser/skus/skus_service_factory.h"
 #include "brave/browser/ui/webui/skus_internals_ui.h"
+#include "brave/components/ai_chat/content/browser/ai_chat_throttle.h"
```
wait. this should also be under `BUILDFLAG`

AlexeyBarabash replied:
oh, my bad, fixing

AlexeyBarabash replied:
I'll build it with `ai_chat_enabled=false` just in case

goodov commented on browser/sources.gni:
```diff
@@ -132,6 +132,7 @@ brave_chrome_browser_deps = [
   "//brave/browser/ui",
   "//brave/browser/ui/tabs",
   "//brave/common",
+  "//brave/components/ai_chat/content/browser",
```
and this under `if (ai_chat_enabled)`

AlexeyBarabash replied:
thanks, fixed at 
https://github.com/brave/brave-core/pull/20996/commits/b46d3237a9a659589182e66c39a9f573030885c4#diff-d00e7b939f2d684a6f82e428ab699596962b739950acc1da8804abb5f607d1b3L135

this dependency  already was at https://github.com/brave/brave-core/blob/master/browser/sources.gni#L289

bridiver commented on components/ai_chat/content/browser/ai_chat_throttle.cc:
```diff
@@ -0,0 +1,79 @@
+    return nullptr;
+  }
+
+  // Purpose of this throttle is to forbid loading of chrome-untrusted://chat
```
Checking the page transition type does not seem like a good way to prevent this. Just get the `WebContents` from the navigation handle and call `sessions::SessionTabHelper::IdForTab`. A non-tab will have `SessionID::InvalidValue`
